offer-manager-domain/src/main/groovy/life/qbic/business/offers/Offer.groovy

```diff
@@ -458,36 +438,25 @@ class Offer {
         return noVatCategory
     }
 
-    /**
-     * Returns a deep copy of all available offer versions.
-     *
-     * @return A list of available offer versions
-     */
-    List<OfferId> getAvailableVersions() {
-        return availableVersions.stream()
-                .map(offerId -> new OfferId(
-                        new RandomPart(offerId.randomPart),
-                        new ProjectPart(offerId.projectPart),
-                        new Version(offerId.version)
-                )).collect()
-    }
-
-    void addAvailableVersion(OfferId offerId) {
-        this.availableVersions.add(new OfferId(offerId))
+    void addAvailableIdentifier(OfferId offerId) {
+        this.availableIdentifiers.add(offerId)
     }
 
-    void addAllAvailableVersions(Collection<OfferId> offerIdCollection) {
-        offerIdCollection.each {addAvailableVersion(it)}
+    void addAllAvailableIdentifiers(Collection<OfferId> offerIdCollection) {
+        offerIdCollection.each {addAvailableIdentifier(it)}
     }
 
     /**
      * Increases the version of an offer, resulting in a offer id with a new version tag.
      */
     void increaseVersion() {
-        def copyIdentifier = new OfferId(this.identifier)
-        identifier = getLatestVersion()
+        def currentId = new OfferId(identifier.projectPart, identifier.randomPart, identifier.version)
+        addAvailableIdentifier(currentId)
+        def latestId = getLatestIdentifier()
+        if (latestId != identifier) {
+            identifier = latestId
+        }
```

Comment:
Imo this part shows, that we expect to much from the Offer object. I think it was a mistake to put all this version handling knowledge into a single object. Here, the state from the actually available offer versions differ already from the persistent application state, which can easily lead to exceptions.

Since you refactored it, I could not just read over it ;) Can you maybe introduce a Set instead of a List for the available OfferIds, such that we dont accidently have duplicate entries in there?

Reply:
Handling Sets with Java 8 and below is a real pain. I checked that no duplicate is entered here and would avoid this in this refactor (as I think that the available versions list here is probably unnecessary)